# nf-sturgeon

[Nextflow] workflow for using [Sturgeon] CNS neural network classifier on Nanopore tumor DNA reads.

Requires either [Docker] or [Apptainer] installed.

Usage:

```sh
nextflow run chusj-pigu/nf-sturgeon [--pod5 INPUT] [OPTIONS]
```

To list available options and parameters for this workflow, run :

``` sh
nextflow run chusj-pigu/nf-sturgeon --help
```

Overview:

This workflow can be run locally or on Compute Canada. To use on Compute Canada, use the `-profile drac` option.

[Sturgeon] uses a high amount of RAM. If using locally, make sure you have access to around 200 GB of RAM, otherwise it will crash.

## 1. Basecalling

Since [Sturgeon] needs modified bases information, basecalling is done with [Dorado] on duplex mode with 5mCG_5hmCG modified base calling, in simplex using the Super accurate algorithm.

## 2. QC stats and filtering

QC plots are generated by [NanoPlot]. Reads that have a QS score <10 will be classified as failed, and subsequent steps are only carried out on reads with QS score >= 10.

## 3. Mapping

Mapping of the reads is done in parallel to GRCh38 and CHM13v2 with [minimap2] using `-ax -map-ont` options for long reads generated by Nanopore sequencing.

## 4. Sorting and indexing

Using [samtools], the .sam files are first converted to .bam using `samtools view`, sorted with `samtools sort` and indexed using `samtools index`. Mapping statistics are computed with [mosdepth].

## 5. Extracting modification scores

The bam files generated by mapping to CHM13v2 are passed to [modkit]. 5hmC calls are converted to 5mC calls as recommended by [Sturgeon] with `modkit adjust-mods`, and scores are extracted with `modkit extract`.

## 6. CNS neural network classification

Bed files are created with `sturgeon inputtobed` and passed to `sturgeon predict` for CNS classification.

## Outputs

Mapping statistics are stored in text files and summarized in a [multiqc] report. [Sturgeon] will output a .bed file and a pdf file showing scores for different CNS tumor types.

[Docker]: https://www.docker.com
[Apptainer]: https://apptainer.org
[Nextflow]: https://www.nextflow.io/docs/latest/index.html
[Dorado]: https://github.com/nanoporetech/dorado
[minimap2]: https://lh3.github.io/minimap2/minimap2.html
[samtools]: http://www.htslib.org
[multiqc]: https://multiqc.info
[mosdepth]: https://github.com/brentp/mosdepth
[NanoPlot]: https://github.com/wdecoster/NanoPlot
[modkit]: https://github.com/nanoporetech/modkit
[Sturgeon]: https://github.com/marcpaga/sturgeon
