# nf-sturgeon

[Nextflow] workflow for using [Sturgeon] CNS neural network classifier on Nanopore tumor DNA reads.

Requires either [Docker] or [Apptainer] installed.

Usage:

```sh
nextflow run chusj-pigu/nf-sturgeon [--pod5 INPUT] [OPTIONS]
```

To list available options and parameters for this workflow, run :

``` sh
nextflow run chusj-pigu/nf-sturgeon --help
```

## Overview:

This workflow can be run locally or on Compute Canada. To use on Compute Canada, use the `-profile drac` option.

[Sturgeon] uses a high amount of RAM. If using locally, make sure you have access to around 200 GB of RAM, otherwise it will crash.

## Inputs

- `--pod5`: Path to the directory containing pod5.
- `--fastq`: Path to the directory containing fastq files. Only use if skipping basecalling.
- `--sample_id`: Will name output files according to sample id [default: reads].
- `--out_dir`: Output directory to place mapped files and reports in [default: output].
- `--ref_hg38`: Path to the hg38 fasta reference genome.
- `--ref_chm13`: Path to the T2T CHm13 fasta reference genome.
- `--bed`: Bed file containing regions of interest to compute mapping statistics with mosdepth.
When running on Digital Research Alliance of Canada Narval:
- `--dorado_model`: Path for the basecalling model, required when running with drac profile [default: path to sup@v5.0.0].
- `--m_bases_path`: Path for the modified basecalling model, required when running with drac profile [default: path to sup@v5.0.0_5mCG_5hmCG].
- `--ubam`: Location of unaligned bam file generated by the unfinished basecalling, usually in the work directory [default: none]

## Outputs

This workflow will output:

| File Path             | Description | Condition        |
| --------------------- | ----------- | ---------------- |
| reads/{sample_id}_passed.fq.gz | Merged raw reads that have passed filter of average QS >= `--minqs` | If --skip_basecall is not used |
| reads/{sample_id}_failed.fq.gz | Merged raw reads that have failed filter of average QS >= `--minqs` | If --skip_basecall is not used |
| alignments/{sample_id}.{ref}.bam<br>{sample_id}.{ref}.bam.bai | Aligned and sorted bam file mapped to reference `--ref` along with it's index | One bam for each reference |
| reports/mosdepth/{sample_id}.{ref}.regions.bed.gz | bed file containing average coverage for each selected region in `--bed` | If `--bed` is provided |
| reports/multiqc_report.html | [multiqc] report containing [Nanoplot] and [mosdepth] outputs | Always |
| sturgeon/merged_probes_methyl_calls.{pdf,bed,csv,txt} | Files containing the sturgeon calls and scores | Always

## Parameters

- `--dorado_model`: Basecalling model to use [default: sup@v5.0.0].
- `--skip_basecall`: Basecalling step will be skipped; input must be in fastq [default: false].
- `--skip_hg38`: Mapping to hg38 will be skipped [default: false].
- `--sturgeon_model`: Local copy of Sturgeon model to use [default for drac: general model]
- `--resume`: Use when basecalling has stopped without finishing, and you want to resume dorado from where it stopped [default: false]
- `--duplex`: Dorado will basecall in duplex mode instead of simplex [default: false].
- `--minqs`: Minimum average read QS score to keep [default: 10]
- `-profile`: Use standard for running locally, test when running a small local test, drac when running on Digital Research Alliance of Canada Narval and test_drac when running tests on Narval [default: standard].
- `--batch`: Batch size for basecalling; if 0, optimal batch size will be automatically selected [default: 0].
- `--dorado_cpu`: Whether to use cpu for basecalling [default: false].

## Steps

## 1. Basecalling

Since [Sturgeon] needs modified bases information, basecalling is done with [Dorado] on duplex mode with 5mC_5hmC modified base calling, in simplex using the Super accurate algorithm.

## 2. QC stats and filtering

QC plots are generated by [NanoPlot]. Reads that have a QS score <10 will be classified as failed, and subsequent steps are only carried out on reads with QS score >= 10.

## 3. Mapping

Mapping of the reads is done in parallel to GRCh38 and CHM13v2 with [minimap2] using `-ax -map-ont` options for long reads generated by Nanopore sequencing. To only map to CHM13v2, use the option `--skip_hg38`.

## 4. Sorting and indexing

Using [samtools], the .sam files are first converted to .bam using `samtools view`, sorted with `samtools sort` and indexed using `samtools index`. Mapping statistics are computed with [mosdepth].

## 5. Extracting modification scores

The bam files generated by mapping to CHM13v2 are passed to [modkit]. 5hmC calls are converted to 5mC calls as recommended by [Sturgeon] with `modkit adjust-mods`, and scores are extracted with `modkit extract`.

## 6. CNS neural network classification

Bed files are created with `sturgeon inputtobed` and passed to `sturgeon predict` for CNS classification.

## Outputs

Mapping statistics are stored in text files and summarized in a [multiqc] report. [Sturgeon] will output a .bed file and a pdf file showing scores for different CNS tumor types.

[Docker]: https://www.docker.com
[Apptainer]: https://apptainer.org
[Nextflow]: https://www.nextflow.io/docs/latest/index.html
[Dorado]: https://github.com/nanoporetech/dorado
[minimap2]: https://lh3.github.io/minimap2/minimap2.html
[samtools]: http://www.htslib.org
[multiqc]: https://multiqc.info
[mosdepth]: https://github.com/brentp/mosdepth
[NanoPlot]: https://github.com/wdecoster/NanoPlot
[modkit]: https://github.com/nanoporetech/modkit
[Sturgeon]: https://github.com/marcpaga/sturgeon
